packages/veritone-redux-common/src/helpers/redux/paginate.js

```diff
@@ -0,0 +1,256 @@
+  // for each engine, so in the LIST_BUILDS paginator we keyBy: 'engineId'.
+
+  // engineId data is expected to live at action.meta[engineId].
+
+  // if keyBy is specified, `key` is ignored. In this example to select from a keyBy'd
+  // reducer, use ie makeSelectors('engineId1'), makeSelectors('engineId2')
+
+  // keyBy can also be a function, receiving (localState, action.meta), to
+  // paginate on other state in the paginated store
+  keyBy = null,
+  // allow the same actions to be used for multiple pagination types by filtering
+  // on something else in the action: (action) => bool
+  filter = constant(true),
+  idAttribute = 'id',
+  selectResponseResults = action =>
+    get(action.payload, 'data.results.records', action.payload.results),
+  selectResponseTotal = action =>
+    get(action.payload, 'data.totalResults.count', action.payload.totalResults)
+}) {
+  if (!Array.isArray(types) || types.length !== 3) {
+    throw new Error('Expected types to be an array of three elements.');
+  }
+
+  if (!types.every(isString)) {
+    throw new Error('Expected types to be strings.');
+  }
+
+  if (!isString(key)) {
+    throw new Error('Expected key to be a string.');
+  }
+
+  if (!isFunction(filter)) {
+    throw new Error('Expected filter to be a function.');
+  }
+
+  if (keyBy && !(isString(keyBy) || isFunction(keyBy))) {
+    throw new Error('Expected keyBy to be a string or function.');
+  }
+
+  const [requestType, successType, failureType] = types;
+
+  let defaultState = {
+    status: fetchingStatus.default,
+    // { 5: 'myEngineId' } where 5 is index of myEngineId in overall server data set
+    // (object because sparse arrays don't work well)
+    idsByPaginationOrder: {},
+    // which IDs are currently fetching (based on offset/limit specified)
+    fetchingEntities: [],
+    failedFetchingEntities: [],
+    totalEntities: 0,
+    // requestId of active request (so new requests can supersede old ones)
+    activeRequestId: null
+  };
+
+  const updatePagination = createReducer(defaultState, {
+    [requestType](state, action) {
+      if (!filter(action)) {
+        return state;
+      }
+
+      const requestSuccessState = {
+        ...state,
+        status: fetchingStatus.fetching,
+        fetchingEntities: range(
+          action.meta.offset,
+          action.meta.offset + action.meta.limit
+        ),
+        failedFetchingEntities: [],
+        activeRequestId: action.meta.requestId
+      };
+
+      return action.error
+        ? // handle requestError ie. offline
+          this[failureType](
+            {
+              ...state,
+              // set needed keys for failure handler
+              ...pick(requestSuccessState, [
+                'fetchingEntities',
+                'activeRequestId'
+              ])
+            },
+            action
+          )
+        : requestSuccessState;
+    },
+
+    [successType](state, action) {
+      if (!filter(action)) {
```

Comment:
any reason you're choosing not to combine these? 

`if (!filter(action) || requestId !== state.activeRequestId)`

Reply:
I guess it's just to avoid complecting two unrelated cases into a single statement